setlexsem/generate/generate_sets.py

```diff
@@ -44,28 +45,69 @@ def get_parser():
     return parser
 
 
+def astype_set(raw_input):
+    if type(raw_input) == str:
+        raw_input = raw_input.strip()
+        set_out = ast.literal_eval(raw_input)
+    else:
+        set_out = raw_input
+    return set_out
+
+
+def parse_set_pair(raw_set_a: str, raw_set_b: str) -> Tuple[set, set]:
+    """Parse string representations of sets into Python sets"""
+    try:
+        set_a = astype_set(raw_set_a)
+        set_b = astype_set(raw_set_b)
+        return set_a, set_b
+    except (ValueError, SyntaxError) as e:
+        logger.error(f"Failed to parse sets: {e}")
+        raise
+
+
+def generate_set_pair(sampler: Union[Iterable, callable]) -> Tuple[set, set]:
+    """Generate a pair of sets from either an iterable or callable sampler"""
+    try:
+        if isinstance(sampler, Iterable):
+            raw_a, raw_b = next(sampler)
+            return parse_set_pair(raw_a, raw_b)
+        else:
+            return sampler()
+    except Exception as e:
+        logger.warning(f"Failed to generate set pair from sampler: {e}")
+        return None, None
+
+
 def make_sets_from_sampler(
     sample_set: Sampler,
     num_runs: int,
 ) -> List[Dict[str, Any]]:
     """Generate random sets from the sampler"""
 
     # initlize the dataset
+    empty_sampler_counter = 0
     set_list = []
     for i in range(num_runs):
-        try:
-            # create two sets from the sampler
-            A, B = sample_set()
-            # loop through operations (on the same random sets)
-            set_list.append(
-                {
-                    "experiment_run": i,
-                    "A": A,
-                    "B": B,
-                }
-            )
-        except:
+        # create two sets from the sampler
+        A, B = generate_set_pair(sample_set)
+        if A is None or B is None:
+            empty_sampler_counter += 1
             continue
+
+        # loop through operations (on the same random sets)
+        set_list.append(
+            {
+                "experiment_run": i,
+                "A": A,
+                "B": B,
+            }
+        )
+
+    if empty_sampler_counter >= 10:
+        logger.warning(
+            f"Sampler is empty {empty_sampler_counter} times out of {num_runs}"
```

Comment:
`f"Sampled at least one empty set {empty_sample_count}  out of {num_runs} samples."